gobblin-salesforce/src/main/java/org/apache/gobblin/salesforce/SalesforceSource.java

```diff
@@ -182,7 +208,183 @@ String generateSpecifiedPartitions(Histogram histogram, int maxPartitions, long
     return specifiedPartitions;
   }
 
-  private Histogram getHistogram(String entity, String watermarkColumn, SourceState state,
+  /**
+   * Compute the target partition size.
+   */
+  private int computeTargetPartitionSize(Histogram histogram, int minTargetPartitionSize, int maxPartitions) {
+     return Math.max(minTargetPartitionSize,
+        DoubleMath.roundToInt((double) histogram.totalRecordCount / maxPartitions, RoundingMode.CEILING));
+  }
+
+  /**
+   * Get a {@link JsonArray} containing the query results
+   */
+  private JsonArray getRecordsForQuery(SalesforceConnector connector, String query) {
+    try {
+      String soqlQuery = SalesforceExtractor.getSoqlUrl(query);
+      List<Command> commands = RestApiConnector.constructGetCommand(connector.getFullUri(soqlQuery));
+      CommandOutput<?, ?> response = connector.getResponse(commands);
+
+      String output;
+      Iterator<String> itr = (Iterator<String>) response.getResults().values().iterator();
+      if (itr.hasNext()) {
+        output = itr.next();
+      } else {
+        throw new DataRecordException("Failed to get data from salesforce; REST response has no output");
+      }
+
+      return GSON.fromJson(output, JsonObject.class).getAsJsonArray("records");
+    } catch (RestApiClientException | RestApiProcessingException | DataRecordException e) {
+      throw new RuntimeException("Fail to get data from salesforce", e);
+    }
+  }
+
+  /**
+   * Get the row count for a time range
+   */
+  private int getCountForRange(TableCountProbingContext probingContext, StrSubstitutor sub,
+      Map<String, String> subValues, long startTime, long endTime) {
+    String startTimeStr = Utils.dateToString(new Date(startTime), SalesforceExtractor.SALESFORCE_TIMESTAMP_FORMAT);
+    String endTimeStr = Utils.dateToString(new Date(endTime), SalesforceExtractor.SALESFORCE_TIMESTAMP_FORMAT);
+
+    subValues.put("start", startTimeStr);
```

Comment:
It's necessary to consider whether we should include the `start` by comparing it with the global partition, which can be put in the `probingContext`